CorsixTH/Lua/research_department.lua

```diff
@@ -505,39 +505,42 @@ function ResearchDepartment:discoverObject(object, automatic)
   self.research_progress[object].discovered = true
 
   -- Go through all rooms to see if another one can be made available.
-  for room, _ in pairs(self.hospital.undiscovered_rooms) do
-    local discovery = true
-    for needed, _ in pairs(room.objects_needed) do
-      local obj = self.research_progress[TheApp.objects[needed]]
-      if obj and not obj.discovered then
-        discovery = false
-        break
-      end
-    end
-    if discovery then
-      self.hospital.discovered_rooms[room] = true
-      self.hospital.undiscovered_rooms[room] = nil
-      if self.hospital == self.world.ui.hospital then
-        if automatic then
-          self.world.ui.adviser:say(_A.research.new_available:format(object.name))
-        else
-          self.world.ui.adviser:say(_A.research.new_machine_researched:format(object.name))
+  local room_discovery = self.hospital.room_discovery
+  for room_id, _ in pairs(room_discovery) do
+    if not room_discovery[room_id].is_discovered then
+      local room = room_discovery[room_id].room
+      local discovery = true
```

Comment:
I know you just moved existing code, but `discovery` is highly generic, can you find a better name?